### PR TITLE
M3-5302: Ability to add a promo code to account

### PR DESCRIPTION
**Pull request by bnussman**
## Description

- Adds ability for users to add promo codes to their accounts

https://user-images.githubusercontent.com/6440455/128060832-bf3f9e9a-cd93-460f-972d-9b048e6687f5.mov

## How to test

- Test adding a promo code
  - Test on accounts with a balance, with credit, etc...
  - Test on restricted user accounts

### CODE REVIEW

**Comment by bnussman** [comment]
Would a success toast be a good idea here?

File: packages/manager/src/features/Billing/BillingPanels/BillingSummary/PromoDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+      })
+      .catch((error) => {
+        setLoading(false);
+        setError(
+          getAPIErrorOrDefault(error, 'Unable to add promo code')[0].reason
+        );
+      });
+  };
+
+  const actions = () => (
```
Comment (HanaXu) [inline packages/manager/src/features/Billing/BillingPanels/BillingSummary/PromoDialog.tsx:57]: This doesn't need to be a function

**Review by HanaXu — COMMENTED** [review]
I don't have an account that's < 90 days so I couldn't check the success state but logically, this looks good and the call to the API was made, error handled, etc. I agree that adding a success toast would be a good idea.

I don't know if this would actually ever happen, but the spacing looks a bit tight if the user has a payment past due:

image

**Comment by bnussman** [comment]
@HanaXu  I don't think the `Add a promo` link should even render if the user has a payment past due according to our internal ticket... Did it do that?

**Comment by HanaXu** [comment]
@bnussman Oh you can disregard that comment then, I was just testing various states by editing the code

File: packages/manager/src/utilities/date.ts [hunk 2]
```diff
@@ -29,3 +31,22 @@ export const isAfter = (d1: string, d2: string) => {
...
+
+/**
+ * Determines whether or not a date has occoured wthin the last
+ * some number of days.
+ * @param days number of days
+ * @param date date to check for. This is a string because it expects
+ * a date as a string returned by the Linode API.
+ * @returns {boolean} true if date has occoured within the last x days
+ */
+export const createdWithinDays = (days: number, date?: string): boolean => {
```
[the remark this answers is not in the document]
Comment (bnussman) [reply packages/manager/src/utilities/date.ts:43]: Added!

**Review by johnwcallahan — APPROVED** [review]
I've got a few minor things, but overall looks good. 

To my eye the spacing looks a _bit_ tight between "Credit" and "Add a promo code". 
Screen Shot 2021-08-05 at 5 06 29 PM

File: packages/manager/src/features/Billing/BillingPanels/BillingSummary/PromoDialog.tsx [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    }
+  }, [open]);
+
+  const addPromo = () => {
+    setLoading(true);
+    setError(undefined);
+    addPromotion(promoCode)
+      .then(() => {
+        setLoading(false);
+        enqueueSnackbar('Successfully added promotion to your account.', {
```
Comment (HanaXu) [inline packages/manager/src/features/Billing/BillingPanels/BillingSummary/PromoDialog.tsx:49]: Nitpick: What about `Successfully applied promo to your account.`? The CTA in the modal was `apply` and we seem to be using the word `promo` over `promotion`

**Review by HanaXu — APPROVED** [review]
LGTM! 💸 💸

**Pull request merged**